File: src/Scripting/PHP/builder.php
<?php

declare(strict_types=1);

interface ReportBuilder
{
    public function reset(): void;
    public function addTitle(string $title): void;
    public function addSection(string $heading, string $body): void;
    public function build(): string;
}

final class TextReportBuilder implements ReportBuilder
{
    /** @var list<string> */
    private array $parts = [];

    public function reset(): void { $this->parts = []; }
    public function addTitle(string $title): void { $this->parts[] = '# ' . $title; }
    public function addSection(string $heading, string $body): void
    {
        $this->parts[] = '## ' . $heading;
        $this->parts[] = $body;
    }
    public function build(): string { return implode("\n", $this->parts); }
}

final class HtmlReportBuilder implements ReportBuilder
{
    /** @var list<string> */
    private array $parts = [];

    public function reset(): void { $this->parts = []; }
    public function addTitle(string $title): void { $this->parts[] = '<h1>' . $title . '</h1>'; }
    public function addSection(string $heading, string $body): void
    {
        $this->parts[] = '<h2>' . $heading . '</h2>';
        $this->parts[] = '<p>' . $body . '</p>';
    }
    public function build(): string { return implode('', $this->parts); }
}

function buildAvailabilityReport(ReportBuilder $builder): string
{
    $builder->reset();
    $builder->addTitle('Service status');
    $builder->addSection('Availability', '99.95%');
    return $builder->build();
}

echo buildAvailabilityReport(new TextReportBuilder()), PHP_EOL;
echo '---', PHP_EOL;
echo buildAvailabilityReport(new HtmlReportBuilder()), PHP_EOL;

File: src/Web/Dart/builder.dart
abstract interface class ReportBuilder {
  void reset();
  void addTitle(String title);
  void addSection(String heading, String body);
  String build();
}

final class TextReportBuilder implements ReportBuilder {
  final List<String> _parts = [];

  @override
  void reset() => _parts.clear();

  @override
  void addTitle(String title) => _parts.add('# $title');

  @override
  void addSection(String heading, String body) =>
      _parts.addAll(['## $heading', body]);

  @override
  String build() => _parts.join('\n');
}

final class HtmlReportBuilder implements ReportBuilder {
  final List<String> _parts = [];

  @override
  void reset() => _parts.clear();

  @override
  void addTitle(String title) => _parts.add('<h1>$title</h1>');

  @override
  void addSection(String heading, String body) =>
      _parts.addAll(['<h2>$heading</h2>', '<p>$body</p>']);

  @override
  String build() => _parts.join();
}

String buildAvailabilityReport(ReportBuilder builder) {
  builder.reset();
  builder.addTitle('Service status');
  builder.addSection('Availability', '99.95%');
  return builder.build();
}

void main() {
  print(buildAvailabilityReport(TextReportBuilder()));
  print('---');
  print(buildAvailabilityReport(HtmlReportBuilder()));
}
